src/pudl/transform/ferc1.py

```diff
@@ -853,6 +853,533 @@ def strip_col_of_non_ints(self, col: pd.Series) -> pd.DataFrame:
+            .pipe(self.associate_notes_with_values)
+            # .pipe(self.drop_invalid_rows)
+        )
+        # Remove headers and note rows
+        df = df[(df["row_type"] != "header") & (df["row_type"] != "note")].copy()
+
+        return df
+
+    def extract_ferc1_license(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Extract FERC license number from plant_name.
+
+        Many FERC license numbers are embedded in the plant_name_ferc1 field, whether
+        as a note or an actual plant name. Not all numbers in the plant_name_ferc1 field
+        that are FERC licenses, however. Some are dates, dollar amounts, page numbers,
+        or numbers of wind turbines. This function extracts valid FERC license numbers
+        and puts them in a new column.
+
+        Valid FERC license numbers are:
+        - Greater than 2.
+        - Usually accompanied by key phrases such as "license", "no.", "ferc", or
+        "project".
+        - Don't contain phrases like "page", "pg", "$", "wind", "units"
+        - Don't fall within the range of a valid year, defined as: 1900-2050
+        - Are categorized as either hydro or NA.
+
+        This function also fills "other" fuel types with hydro for all these plants.
+        """
+        logger.info(f"{self.table_id.value}: Extracting FERC license from plant name")
+        # Extract all numbers greater than 2 digits from plant_name_ferc1 and put them
+        # in a new column as integers.
+        out_df = df.assign(
+            license_id_ferc1=lambda x: (
+                x.plant_name_ferc1.str.extract(r"(\d{3,})")
+                .astype("float")
+                .astype("Int64")
+            ),
+        )
+        # Define what makes a good license
+        obvious_license = out_df.plant_name_ferc1.str.contains(
+            r"no\.|license|ferc|project", regex=True
+        )
+        not_license = out_df.plant_name_ferc1.str.contains(
+            r"page|pg|\$|wind|solar|nuclear|nonutility|units|surrendered", regex=True
+        )
+        exceptions = out_df.plant_name_ferc1.str.contains(
+            r"tomahawk|otter rapids|wausau|alexander|hooksett|north umpqua", regex=True
+        )
+        year_vs_num = (out_df["license_id_ferc1"] > 1900) & (
+            out_df["license_id_ferc1"] < 2050
+        )
+        not_hydro = ~out_df["plant_type"].isin(["hydro", np.nan, None]) | ~out_df[
+            "fuel_type"
+        ].isin(["hydro", "other"])
+        # Replace all the non-license numbers with NA
+        out_df.loc[
+            (not_hydro & ~obvious_license)
+            | not_license
+            | (year_vs_num & ~obvious_license & ~exceptions),
+            "license_id_ferc1",
+        ] = np.nan
+        # Fill fuel type with hydro
+        out_df.loc[
+            out_df["license_id_ferc1"].notna() & (out_df["fuel_type"] == "other"),
+            "fuel_type",
+        ] = "hydro"
+
+        return out_df
+
+    def label_row_type(self, df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
Right now, this particular function has nested functions inside of it. I'm not sure how folks feel about this. I can certainly peel them apart and have a `.pipe(self.label_total_rows).pipe(self.label_header_rows).pipe(self.label_note_rows)` instead of just the one `.pipe(self.label_row_type)`

Reply:
the more i think about this the more i think the nesting is fine!.. I think looking at this in the github web interface has been hard but looking at it in a text editor really changed my mind (bc collapsing)

Reply:
Done!